### PR TITLE
Harden the MatmulTensorCore strategy

### PR DESCRIPTION
**Pull request by nicolasvasilache**
This revision improves the behavior of the matmul tensor core strategy for aligned and unaligned cases as follows:
- provide better behavior around using the maximum threads for copies by avoiding rank-reducing patterns on vectors that trip up late vector-to-mma conversions.
- plug the barrier removal pass added in #13886 : this removes an unnecessary barrier between copies to shared memory and allows a form that the mma.sync pipeline accepts without choking.
- force the synchronization of the sequential loop before hoisting redundant transfers to workaround issue #13964. This addresses synchronization issues reported in #13451 and #13811.
- various NFC cleanups and error message improvements.
- addition of failsafe configuration alterations to avoid hitting corner cases where we would fail to tile and generate sequential code.
- turning on mma.sync by default in the unaligned and partially aligned cases.

The changes have been tested against other implementations for numerical accuracy and against races in a variety of unaligned/partially aligned and aligned [cases](https://github.com/iree-org/iree-samples/blob/f4ea2d5e5d0c0514a3d88a1926ffa6fc2d721e9f/transform_dialect/python/matmul_test.py) with various compiler configurations.

### CODE REVIEW

**Review by matthias-springer — APPROVED** [review]

File: compiler/src/iree/compiler/Codegen/TransformDialectStrategies/GPU/Common.cpp [hunk 1]
```diff
@@ -549,6 +551,15 @@ Value mlir::iree_compiler::gpu::buildConvertToTensorCoreOp(
...
     b.create<ApplyPatternsOp>(funcH, config);
   }
+
+  Value forH = b.create<transform::MatchOp>(
+      transform::OperationType::get(b.getContext(), "scf.for"), funcH,
+      b.getStrArrayAttr({scf::ForOp::getOperationName()}),
+      /*matchInterfaceEnum=*/transform::MatchInterfaceEnumAttr(),
+      /*opAttrs=*/DictionaryAttr(),
+      /*filterResultType=*/TypeAttr());
+  b.create<SynchronizeLoopOp>(forH);
```
Comment (qcolombet) [inline compiler/src/iree/compiler/Codegen/TransformDialectStrategies/GPU/Common.cpp]: This transform is required for correctness, right?
How do we make sure we don't forget about it if we were to do a similar transform somewhere else?

Put differently, shouldn't the barrier be introduced by the optimization that needs it in the first place?

Then we can just keep the later `EliminateGpuBarriersOp`.

**Comment by nicolasvasilache** [comment]
Interestingly, the UI does not surface the gains, so I am pasting manually:

Before this PR @ #14052 ([link](https://gist.github.com/iree-github-actions-bot/657d0528ab9d4e3cf45b3b9834e1e8d2)):


Benchmark Name | Average Latency (ms) | Median Latency (ms) | Latency Standard Deviation (ms)
-- | -- | -- | --
matmul\_123x2561x2561\_f32t\_f32t\_f32t\_tile\_config\_default(linalg) [cuda-sm\_80-linux\_gnu-cuda][ukernel,matmul] cuda(none)[full-inference,default-flags] with zeros @ a2-highgpu-1g[gpu] | 293.722 | 293.654 | 0.231
matmul\_2562x2561x2561\_f32t\_f32t\_f32t\_tile\_config\_default(linalg) [cuda-sm\_80-linux\_gnu-cuda][ukernel,matmul] cuda(none)[full-inference,default-flags] with zeros @ a2-highgpu-1g[gpu] | 1.443 | 1.440 | 0.010
matmul\_2562x2564x2562\_f32t\_f32t\_f32t\_tile\_config\_default(linalg) [cuda-sm\_80-linux\_gnu-cuda][ukernel,matmul] cuda(none)[full-inference,default-flags] with zeros @ a2-highgpu-1g[gpu] | 1.581 | 1.581 | 0.002
matmul\_2564x2564x2564\_f32t\_f32t\_f32t\_tile\_config\_default(linalg) [cuda-sm\_80-linux\_gnu-cuda][ukernel,matmul] cuda(none)[full-inference,default-flags] with zeros @ a2-highgpu-1g[gpu] | 1.187 | 1.187 | 0.000


With this PR ([link](https://gist.github.com/iree-github-actions-bot/bc494defbe7021734526366752cba9be)):

Benchmark Name | Average Latency (ms) | Median Latency (ms) | Latency Standard Deviation (ms)
-- | -- | -- | --
matmul\_123x2561x2561\_f32t\_f32t\_f32t\_tile\_config\_default(linalg) [cuda-sm\_80-linux\_gnu-cuda][ukernel,matmul] cuda(none)[full-inference,default-flags] with zeros @ a2-highgpu-1g[gpu] | 0.197 | 0.197 | 0.000
matmul\_2562x2561x2561\_f32t\_f32t\_f32t\_tile\_config\_default(linalg) [cuda-sm\_80-linux\_gnu-cuda][ukernel,matmul] cuda(none)[full-inference,default-flags] with zeros @ a2-highgpu-1g[gpu] | 1.350 | 1.350 | 0.000
matmul\_2562x2564x2562\_f32t\_f32t\_f32t\_tile\_config\_default(linalg) [cuda-sm\_80-linux\_gnu-cuda][ukernel,matmul] cuda(none)[full-inference,default-flags] with zeros @ a2-highgpu-1g[gpu] | 1.371 | 1.371 | 0.000
matmul\_2564x2564x2564\_f32t\_f32t\_f32t\_tile\_config\_default(linalg) [cuda-sm\_80-linux\_gnu-cuda][ukernel,matmul] cuda(none)[full-inference,default-flags] with zeros @ a2-highgpu-1g[gpu] | 1.081 | 1.077 | 0.012

File: compiler/src/iree/compiler/Codegen/TransformDialectStrategies/GPU/Common.cpp (hunk 1 above)
Comment (nicolasvasilache) [reply compiler/src/iree/compiler/Codegen/TransformDialectStrategies/GPU/Common.cpp]: Great point, I'll create a specific IREE transform to perform this and wrap in barriers.
This requires an upstream refactoring that will take time to refactor so just adding a TODO in this PR for now.
Once integrate has gone through I can finalize.

File: compiler/src/iree/compiler/Codegen/TransformStrategies/GPU/MatmulTensorCoreStrategy.h [hunk 2]
```diff
@@ -74,62 +95,170 @@ class MatmulStrategy : public AbstractGemmLikeStrategy {
...
-    assert(blockTileN() % numThreadsN == 0 &&
-           "blockTileSizes[1] must be divisible by numThreadsN");
+    int64_t numThreadsN = mlir::ceilDiv(blockTileN(), resCopyVectorSize());
+    int64_t numThreadsM =
+        std::min(blockTileM(), mlir::ceilDiv(totalNumThreads(), numThreadsN));
     return MappingInfo{/*numThreads=*/{numThreadsM, numThreadsN},
                        /*tileSizes=*/
-                       {blockTileM() / numThreadsM, blockTileN() / numThreadsN},
+                       {std::max(1l, mlir::ceilDiv(blockTileM(), numThreadsM)),
+                        std::max(1l, mlir::ceilDiv(blockTileN(), numThreadsN))},
```
Comment (ScottTodd) [inline compiler/src/iree/compiler/Codegen/TransformStrategies/GPU/MatmulTensorCoreStrategy.h:152]: This fails to build on Windows and macOS:
https://github.com/openxla/iree/actions/runs/5253473763/jobs/9490854543#step:6:6769
https://github.com/openxla/iree/actions/runs/5253473763/jobs/9490854202#step:8:7566

```
In file included from /Users/antiagainst/actions-runner/_work/iree/iree/compiler/src/iree/compiler/../../iree/compiler/Codegen/TransformStrategies/GPU/Strategies.h:10:
/Users/antiagainst/actions-runner/_work/iree/iree/compiler/src/iree/compiler/../../iree/compiler/Codegen/TransformStrategies/GPU/MatmulTensorCoreStrategy.h:151:25: error: no matching function for call to 'max'
                       {std::max(1l, mlir::ceilDiv(blockTileM(), numThreadsM)),
                        ^~~~~~~~
/Library/Developer/CommandLineTools/SDKs/MacOSX13.1.sdk/usr/include/c++/v1/__algorithm/max.h:39:1: note: candidate template ignored: deduced conflicting types for parameter '_Tp' ('long' vs. 'long long')
max(const _Tp& __a, const _Tp& __b)
^
/Library/Developer/CommandLineTools/SDKs/MacOSX13.1.sdk/usr/include/c++/v1/__algorithm/max.h:50:1: note: candidate template ignored: could not match 'initializer_list<type-parameter-0-0>' against 'long'
max(initializer_list<_Tp> __t, _Compare __comp)
^
/Library/Developer/CommandLineTools/SDKs/MacOSX13.1.sdk/usr/include/c++/v1/__algorithm/max.h:59:1: note: candidate function template not viable: requires single argument '__t', but 2 arguments were provided
max(initializer_list<_Tp> __t)
^
/Library/Developer/CommandLineTools/SDKs/MacOSX13.1.sdk/usr/include/c++/v1/__algorithm/max.h:30:1: note: candidate function template not viable: requires 3 arguments, but 2 were provided
max(const _Tp& __a, const _Tp& __b, _Compare __comp)
^
```

Comment (ScottTodd) [reply compiler/src/iree/compiler/Codegen/TransformStrategies/GPU/MatmulTensorCoreStrategy.h:152]: Fixed in https://github.com/openxla/iree/commit/bd80c9aa59ac78ad3126c467a26f44f5a6e7f90d

**Pull request merged**